kernel_module resource: added blacklisting, enabled, disabled, docs and unit tests

aaronlippold opened the pull request:
Fixes #1643
Fixes #1673 

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Fix up methods, add command mock, do string matching in ruby instead of command

Fixes #1643
Fixes #1673

Signed-off-by: Aaron Lippold <lippold@gmail.com>

adamleff commented on lib/resources/kernel_module.rb:
```diff
@@ -37,13 +66,56 @@ def loaded?
+
+    def blacklisted?
+      if inspec.os.redhat? || inspec.os.name == 'fedora'
+        modprobe_cmd = "/sbin/modprobe --showconfig | grep blacklist | grep #{@module}"
```
It's a bit repetitious to be checking the OS in each of these methods. Can you create a method like:

```
def cmd_for_os
  if inspec.os.redhat? || inspec.os.name == 'fedora'
    '/sbin/modprobe'
  else
    'modprobe'
  end
end
```

And then use that in each of these methods?

```
modprobe_cmd = "#{cmd_for_os} --showconfig | grep blacklist | grep #{@module}"
```

aaronlippold replied:
Sure that is fine.

adamleff commented on lib/resources/kernel_module.rb:
```diff
@@ -37,13 +66,56 @@ def loaded?
+        modprobe_cmd = "modprobe --showconfig | grep blacklist | grep #{@module}"
+      end
+      cmd = inspec.command(modprobe_cmd)
+      cmd.exit_status.zero? ? cmd.stdout.delete("\n") : nil
```
Can you explain what you're trying to do with this line (and ones similar to it)?

Usually, Ruby methods that end in a `?` are expected to return a boolean value. But here, you'll either be returning the stdout without "\n" characters, or nil.

aaronlippold replied:
Sure, I can drop the `?`. I can see that convention? Perhaps with the `*_via_bin_{true|false}` methods should actually return a true or false or even perhaps all of them.

Does it make more sense to change the `cmd.stdout.delete("\n")` lines to return `true` rather than the standard out? Logically inspec would interrupt the `should be_*` and `should_not be_*` functions as a true/false value anyway correct? ( @arlimus I am thinking correctly on this? ) 

What do you think?

adamleff replied:
I think you're misunderstanding. The method name is fine and expected. The ? character in the last line is the Ruby ternary operator and, as written, will not return true or false as is expected for a `be_` matcher or a method that is expected to return Boolean. This is what needs to be corrected.

aaronlippold replied:
@adamleff yes I get that. Sorry I was a little stream of consciousness there. I was asking if you thought that I should return `true` and `nil` in these methods - given that is the logic flow - rather than a string of the `stdout` or `nil`. That might be a better implementation. We are asking the function to return an "answer" but we don't care about the data.

aaronlippold replied:
So, is this a better approach?
```
cmd = inspec.command(modprobe_cmd)
cmd.exit_status.zero? ? true: nil
```
or, to be fully within the expected behaviour of the `?` convention.
```
cmd = inspec.command(modprobe_cmd)
cmd.exit_status.zero? ? true: false
```
i.e. if the command completed successfully, then return true, else return nil.

adamleff replied:
It should return either `true` or `false`

adamleff replied:
And since `cmd.exit_status.zero?` already returns true or false, there is no need for a ternary operator here. Just return the output of that.

adamleff commented on docs/resources/kernel_module.md.erb:
```diff
@@ -10,14 +10,48 @@ Use the `kernel_module` InSpec audit resource to test kernel modules on Linux pl
+* `{ should be_enabled }` is an alias to `loaded` to keep consistant language with `disabled`
+* `{ should be_disabled }` tests if the module is disabled via a fake install using /bin/false or /bin/true
+* `{ should be_disabled_via_bin_false }` tests if the module is disabled via a fake install using /bin/false
+* `{ should be_disabled_via_bin_ture }` tests if the module is disabled via a fake install using /bin/true
```
Typo: `ture` vs `true`

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -11,18 +11,48 @@
+    _(resource.disabled?).must_equal false
+  end
+
+  it 'verify kernel_module parsing disabled' do
```
This looks like a duplicate test to the one in lines 29-32, just with a different module name? We can remove this.

aaronlippold replied:
Ok. Will do.

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -11,18 +11,48 @@
+    _(resource.disabled?).must_equal false
+  end
+
+  it 'verify kernel_module disabled via /bin/true' do
```
This is a good test to make sure that the `disabled_via_bin_true?` method returns false for an unlisted module, but we still need a test for a listed module.

I'd rename this test to "verify an unlisted kernel_module is not disabled via /bin/true" and leave the rest of it alone, and then add another test using this existing title ("verify kernel_module disabled via /bin/true") and test module `nvidiafb` and checking that the method returns `true`

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -11,18 +11,48 @@
+    _(resource.disabled_via_bin_true?).must_equal false
+  end
+
+  it 'verify kernel_module disabled via /bin/false' do
```
This is a good test to make sure that the `disabled_via_bin_false?` method returns false for an unlisted module, but we still need a test for a listed module.

I'd rename this test to "verify an unlisted kernel_module is not disabled via /bin/true" and leave the rest of it alone, and then add another test using this existing title ("verify kernel_module disabled via /bin/true") and test module `sstfb` and checking that the method returns `true`

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -11,18 +11,48 @@
     _(resource.version).must_equal '3.2.2'
   end
+
+  it 'verify kernel_module blacklisting' do
```
We should have a negative test here, too... "verify that an unlisted kernel_module is not blacklisted" and provide a module name that's not in your mocked data and ensure the method returns false.

adamleff reviewed (changes requested):
Super close @aaronlippold - just need a few additional tests to properly test your new functionality.

aaronlippold commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -6,23 +6,119 @@
+  end
+
+  #@todo shouldn't an unlisted module be considered disabled? - return true
+  it 'Verify an unlisted kernel_module is not `disabled` - false' do
```
@adamleff @arlimus @chris-rock RFC: An unlisted kernel module be considered 'disabled' yes? i.e. if an unlisted module does not return true to a 'disabled' query, then we should document the logic and or thinking and or update the code to cover this logic.

adamleff commented on docs/resources/kernel_module.md.erb:
```diff
@@ -10,14 +10,52 @@ Use the `kernel_module` InSpec audit resource to test kernel modules on Linux pl
-    describe kernel_module('module_name') do
-      it { should be_loaded }
-    end
+  ddescribe kernel_module('video') do
```
Typo: `ddescribe` instead of `describe`

adamleff reviewed (approved):
One small docs typo, but otherwise approved.

aaronlippold commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -6,23 +6,118 @@
-  it 'verify kernel_module version' do
-    resource = load_resource('kernel_module', 'dhcp')
-    _(resource.version).must_equal '3.2.2'
+  it 'Verify an unlisted kernel_module is not `enabled` - false' do
```
Should this not return true, if the module is not listed, then it is not enabled - i.e. it is not loaded.

aaronlippold replied:
Basically it is a logic flow question, i.e. an unlisted mod comes back false as disabled. But given that it is not loaded shouldn't it come back as true for 'disabled' - even though it is a kind of 'soft disable'

adamleff commented on lib/resources/kernel_module.rb:
```diff
@@ -37,19 +69,56 @@ def loaded?
-      else
-        modinfo_cmd = "modinfo -F version #{@module}"
-      end
+    # @note maintain logical agreement with 'disabled?' method
```
This line and the one after it should just be removed rather than commented out.

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -6,23 +6,118 @@
 
-  it 'verify kernel_module parsing' do
-    resource = load_resource('kernel_module', 'dhcp')
+  #enabled
```
These next lines should be removed instead of commented out.

adamleff commented on test/unit/resources/kernel_module_test.rb:
```diff
@@ -6,23 +6,118 @@
-  it 'verify kernel_module version' do
-    resource = load_resource('kernel_module', 'dhcp')
-    _(resource.version).must_equal '3.2.2'
+  # it 'Verify an unlisted kernel_module is not `enabled` - false' do
```
These lines should be removed instead of commented out.

adamleff reviewed (changes requested):
@aaronlippold really close. The lines that you commented out that I pointed out need to be removed for cleanliness instead of left in commented out. It will be confusing to the next contributor to wonder why these are commented out. Once this is done and we have a clean single-commit PR, it has my approval.

aaronlippold reviewed (commented):
requested changes pushed

adamleff reviewed (approved):
This looks good now, @aaronlippold  - thanks.

chris-rock commented on docs/resources/kernel_module.md.erb:
```diff
@@ -10,14 +10,49 @@ Use the `kernel_module` InSpec audit resource to test kernel modules on Linux pl
-    describe kernel_module('module_name') do
-      it { should be_loaded }
-    end
+  describe kernel_module('video') do
```
I think we should keep the generic part here and move the examples into the examples sections to keep consistent with other docs

chris-rock commented on lib/resources/kernel_module.rb:
```diff
@@ -36,19 +68,51 @@ def loaded?
+      end
+    end
+
+    def modprobe_cmd_for_os
```
👍

chris-rock commented on lib/resources/kernel_module.rb:
```diff
@@ -36,19 +68,51 @@ def loaded?
+    end
 
-      cmd = inspec.command(modinfo_cmd)
+    def disabled_via_bin_true?
```
Instead of using `disabled_via_bin_true?` I would prefer to offer the a `install_command` and `remove_command` method to return the executed command of a kernel module.

aaronlippold replied:
So your thinking we just add bin_true and bin_false to the blacklisted? method and use a case statement?

chris-rock commented on lib/resources/kernel_module.rb:
```diff
@@ -36,19 +68,51 @@ def loaded?
+      !modprobe_output.match(%r{^install\s+#{@module}\s+/(s?)bin/false}).nil?
+    end
+
+    def blacklisted?
```
Is checking for blacklist enough? A module can be disabled via `/bin/true` too. We could make `disabled_via_bin_false` and `disabled_via_bin_true?` private and add the check to `blacklisted?`.

aaronlippold replied:
So I am returning a true of false in each of those, thus if I put the result of each of them into an array and then checked if the array contained true then return true else false. 

That makes sense?

aaronlippold replied:
Like this?
```
    def blacklisted?
      results = []
      results.push(!modprobe_output.match(/^blacklist\s+#{@module}/).nil?)
      results.push(disabled_via_bin_true?)
      results.push(disabled_via_bin_false?)
      results.uniq!.include?(true) ? true : false
    end
```

chris-rock replied:
I was thinking more about
```
!modprobe_output.match(/^blacklist\s+#{@module}/).nil? || disabled_via_bin_true? || disabled_via_bin_false?
```

aaronlippold replied:
Yes I was thinking that as well at first then I started to overthink it and did the above approach, but your right ( as was I ) an 'or' clause produces the same effect - i.e. return true if any of these return true else return false. ( :smack head ) 

Pushed an update to use the cleaner approach. 

Everything should be good to go.

chris-rock reviewed (changes requested):
@aaronlippold Awesome to see blacklisting checks coming to the resource. While this works, I'd like to do another small round of improvements.

We can also turn `disabled_via_bin_false?` and `disabled_via_bin_true?` private and call them from `blacklisted?` to get this PR merged and add further improvements in another PR.

aaronlippold committed:
Merge branch 'master' into al/kernel_module_blacklist

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
updated resource to group 'bin/false' and 'bin/true' into `blacklisted?`
moved `bin_true?` and `bin_false?` into private methods
updated documentation
udpated unit tests

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
updated to use cleaner apprach to the `blacklisted?` method.

Signed-off-by: Aaron Lippold <lippold@gmail.com>

chris-rock commented on docs/resources/kernel_module.md.erb:
```diff
@@ -4,20 +4,62 @@ title: About the kernel_module Resource
 
-A `kernel_module` resource block declares a module name, and then tests if that module is a loadable kernel module:
+A `kernel_module` resource block declares a module name, and then tests if that
+module is a loadable kernel module:
```
As with the other examples, we need to have only one example here that explains the resource. All other parts should move at the end of the file under examples section.

chris-rock commented on lib/resources/kernel_module.rb:
```diff
@@ -1,20 +1,62 @@
+      it { should be_loaded }
+    end
+
+    # via 'bin_false'
```
I think it does not matter if it was blacklisted by `/bin/false`, `/bin/true` or blacklist config, lets simplify the examples

chris-rock commented on docs/resources/kernel_module.md.erb:
```diff
@@ -4,20 +4,62 @@ title: About the kernel_module Resource
       it { should be_loaded }
     end
 
+    # via 'bin_false'
```
I think it does not matter if it was blacklisted by `/bin/false`, `/bin/true` or blacklist config, lets simplify the examples

chris-rock commented on lib/resources/kernel_module.rb:
```diff
@@ -1,20 +1,62 @@
+
+    describe kernel_module('video') do
+      it { should be_loaded }
+      it { should_not be_disabled }
```
Should we add a `def enabled?`?

adamleff replied:
I'm 👎  on that. I'd like us to pick one (I'd prefer `disabled?`) and then use `should` or `should_not` accordingly.

chris-rock replied:
Perfect, no additional work then :-)

chris-rock reviewed (changes requested):
We are so close @aaronlippold

aaronlippold committed:
Made requested changes to documentation, moved examples and simplified main examples.

Signed-off-by: Aaron Lippold <lippold@gmail.com>

chris-rock reviewed (approved):
Thank you @aaronlippold
